File: app/src/main/java/me/proteus/myeye/io/SpeechDecoderResult.java
package me.proteus.myeye.io;

import com.google.gson.*;

import java.util.ArrayList;
import java.util.Map;

public class SpeechDecoderResult {

    public float confidence;
    public float start;
    public float end;
    public String word;


    public SpeechDecoderResult(float c, float s, float e, String w) {
        this.confidence = c;
        this.start = s;
        this.end = e;
        this.word = w;
    }

    public static ArrayList<SpeechDecoderResult> deserialize(String json) {

        ArrayList<SpeechDecoderResult> list = new ArrayList<>();
        JsonObject obj = JsonParser.parseString(json).getAsJsonObject();

        if (obj.has("alternatives")) {

            JsonArray alternatives = obj.getAsJsonArray("alternatives");
            for (JsonElement el : alternatives) {
                ArrayList<SpeechDecoderResult> part = getSingleResult(el.getAsJsonObject());
                list.addAll(part);
            }

        }

        if (obj.has("result")) {

            JsonArray resultArray = obj.getAsJsonArray("result");
            for (JsonElement el : resultArray) {

                JsonObject result = el.getAsJsonObject();

                list.add(new SpeechDecoderResult(
                        result.get("conf").getAsFloat(),
                        result.get("start").getAsFloat(),
                        result.get("end").getAsFloat(),
                        result.get("word").getAsString()
                ));

            }
        }

        return list;

    }

    private static ArrayList<SpeechDecoderResult> getSingleResult(JsonObject json) {

        ArrayList<SpeechDecoderResult> list = new ArrayList<>();

//        System.out.println("JSON: " + json);

        if (json.has("confidence")) {

            if (json.has("result")) {

                JsonArray resultArray = json.getAsJsonArray("result");
                float avgConf = json.get("confidence").getAsFloat() / resultArray.size();

                for (JsonElement el : resultArray) {

                    JsonObject obj = el.getAsJsonObject();

                    list.add(new SpeechDecoderResult(
                            avgConf,
                            obj.get("start").getAsFloat(),
                            obj.get("end").getAsFloat(),
                            obj.get("word").getAsString()
                    ));
                }
            }

        } else {
            list.add(new SpeechDecoderResult(
                json.get("conf").getAsFloat(),
                json.get("start").getAsFloat(),
                json.get("end").getAsFloat(),
                json.get("word").getAsString()
            ));
        }

        return list;

    }

}

File: composeApp/src/commonMain/kotlin/me/proteus/myeye/io/SpeechDecoderResult.kt
package me.proteus.myeye.io

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.float
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

@Serializable
data class SpeechDecoderResult(
    val confidence: Float,
    val start: Float,
    val end: Float,
    val word: String
)

fun deserialize(json: String): List<SpeechDecoderResult> {
    val list = mutableListOf<SpeechDecoderResult>()
    val obj = Json.parseToJsonElement(json).jsonObject

    obj["alternatives"]?.jsonArray?.forEach { el ->
        list.addAll(getSingleResult(el.jsonObject))
    }

    obj["result"]?.jsonArray?.forEach { el ->
        val result = el.jsonObject
        list.add(SpeechDecoderResult(
            result["conf"]!!.jsonPrimitive.float,
            result["start"]!!.jsonPrimitive.float,
            result["end"]!!.jsonPrimitive.float,
            result["word"]!!.jsonPrimitive.content
        ))
    }

    return list
}

private fun getSingleResult(json: JsonObject): List<SpeechDecoderResult> {
    val list = mutableListOf<SpeechDecoderResult>()

    if (json.containsKey("confidence")) {
        val resultArray = json["result"]!!.jsonArray
        val avgConf = json["confidence"]!!.jsonPrimitive.float / resultArray.size
        resultArray.forEach { el ->
            val obj = el.jsonObject
            list.add(SpeechDecoderResult(
                avgConf,
                obj["start"]!!.jsonPrimitive.float,
                obj["end"]!!.jsonPrimitive.float,
                obj["word"]!!.jsonPrimitive.content
            ))
        }
    } else {
        list.add(SpeechDecoderResult(
            json["conf"]!!.jsonPrimitive.float,
            json["start"]!!.jsonPrimitive.float,
            json["end"]!!.jsonPrimitive.float,
            json["word"]!!.jsonPrimitive.content
        ))
    }

    return list
}
